### PR TITLE
Fix per experiment irt regression

### PR DESCRIPTION
**Pull request by JemmaLDaniel**
# Per-experiment RT-to-iRT linear regression

## Motivation

The existing `RetentionTimeFeature` uses a single `MLPRegressor` to map observed retention times (RT) to indexed retention times (iRT) across the entire dataset. This is problematic for two reasons:

1. **The RT-to-iRT mapping is experiment-specific.** Different LC-MS experiments have different chromatographic conditions (column, gradient, temperature, etc.), so a single global regressor conflates distinct linear relationships. When training or predicting on multi-experiment data, the MLP fits an average mapping that is suboptimal for every individual experiment.

2. **The relationship is linear.** RT and iRT are related by a simple affine transform within a single experiment. An MLP is unnecessarily complex for this; it introduces 7 hyperparameters (`hidden_dim`, `learning_rate_init`, `alpha`, `max_iter`, `early_stopping`, `validation_fraction`, `seed`), risks overfitting on small experiments, and adds nondeterminism from stochastic optimisation. A `LinearRegression` is the natural fit.

A secondary issue is that the regressor was only fitted during training. While the trained MLP was pickled with the calibrator model and therefore available at inference time, it carried the training data's RT-to-iRT mapping. This will frequently be wrong for inference data from entirely different experiments with different chromatographic conditions. The regressor should always be re-fitted from the current data. The `compute-features` entrypoint explicitly blocked unlabelled data with `RetentionTimeFeature` because of this `if labelled` gate.

### Per-experiment linear regressors (`calibration_features.py`)

- Replace the single `MLPRegressor` with a `Dict[str, LinearRegression]` (`irt_predictors`), keyed by experiment name.
- `prepare()` groups spectra by `experiment_name`, selects the top `train_fraction` by confidence per experiment, batches all Koina iRT calls into a single request, then fits one `LinearRegression` per experiment.
- If `experiment_name` is absent, falls back to a single `__global__` regressor with a warning.
- Experiments that already have a fitted regressor (e.g. loaded from a checkpoint) are skipped.
- `compute()` applies the correct per-experiment regressor when predicting iRT from observed RT.
- New `_select_training_data()` helper with a `min_train_points` guard that raises early with a clear error message if an experiment has too few valid spectra.

### Always call `prepare()` (`calibrator.py`)

- Remove the `if labelled` gate on `feature.prepare()`. The RT regressor is self-supervised (uses high-confidence de novo predictions, not database labels), so it must run at both training and inference time.
- This also removes the `compute-features` entrypoint restriction that blocked `RetentionTimeFeature` on unlabelled data.

### Regressor checkpoint workflow (`main.py`, configs)

By default, the RT-to-iRT regressor is re-fitted from the current data at both training and inference time. This is the right behaviour for the general pretrained-model workflow where inference data comes from unseen experiments. However, there is an important within-experiment use case that requires carrying regressors forward: when a calibrator is trained on the subset of spectra that received database search labels and then applied to the remaining unlabelled de novo predictions from the same experiment(s). The unlabelled portion is typically contains a greater proportion of lower-quality spectra (the ones the database search couldn't confidently match), and its de novo prediction confidence distribution is likely to be skewed towards lower-confidence predictions. Fitting the RT-to-iRT regressor from the top predictions in this skewed distribution produces a noisier fit than one derived from the higher-confidence labelled portion. By saving the regressors fitted during training and loading them at inference time, the calibrator uses the cleaner mapping from the labelled data for experiments it has already seen, while still fitting fresh regressors for any new experiments encountered at inference.

- `save_regressors()` / `load_regressors()` methods on `RetentionTimeFeature` for persisting per-experiment regressors to a pickle file, separate from the calibrator model.
- `__getstate__` / `__setstate__` exclude transient regressor state from the calibrator pickle — regressors are always re-fitted from data unless explicitly loaded.
- `train.yaml`: new `irt_regressor_output_path` option to save regressors after training.
- `predict.yaml`: new `irt_regressor_path` option to load pre-fitted regressors at inference time.
- `main.py`: `train_entry_point` saves regressors when configured; `predict_entry_point` loads them when configured.

### `experiment_name` always available (`data_loaders.py`)

- `InstaNovoDatasetLoader` and `MZTabDatasetLoader` now derive `experiment_name` from the file stem when the column is not already present, even when `add_index_cols=False`. This ensures the per-experiment grouping works without requiring explicit user configuration.

### Configuration simplification (`calibrator.yaml`)

- Remove 7 MLP hyperparameters (`hidden_dim`, `learning_rate_init`, `alpha`, `max_iter`, `early_stopping`, `validation_fraction` and the MLP-specific `seed` description).
- Add `min_train_points` (default 10).
- `train_fraction` and `seed` retained with updated descriptions.

### Tests

- Rewrite `RetentionTimeFeature` tests to use pre-fitted `LinearRegression` instead of mocking `MLPRegressor.predict`.
- Add `test_pickle_excludes_regressor_state` and `test_save_and_load_regressors`.
- Update `test_data_loaders.py` to expect `experiment_name` always present.
- Update `learn_from_missing` filtering tests for the new constructor signature.

### Documentation

- Rewrite `RetentionTimeFeature` section in `docs/api/calibration.md` covering per-experiment fitting, the `experiment_name` column, configuration parameters, and the regressor checkpoint workflow.
- Update `docs/configuration.md` with the simplified config block and remove the `labelled` restriction note.

### CODE REVIEW

File: winnow/calibration/calibration_features.py [hunk 1]
```diff
@@ -1200,65 +1186,116 @@ def check_valid_irt_prediction(self, dataset: CalibrationDataset) -> pd.Series:
...
+                    "MGF format (which derives it from the filename).",
+                    stacklevel=2,
+                )
+                experiments_to_fit = {"__global__": dataset.metadata}
+            else:
+                return
+        else:
+            experiments_to_fit = {
+                str(exp_name): group
+                for exp_name, group in dataset.metadata.groupby("experiment_name")
```
Comment (BioGeek) [inline winnow/calibration/calibration_features.py]: If an input file already has a non-string `experiment_name` column, `prepare()` stores the regressor under `str(exp_name)`, but `compute() ` later looks up `self.irt_predictors[exp_name]` using the original value, causing a  `KeyError` for integer/categorical experiment IDs. The loaders document that existing `experiment_name` columns are used as-is, so either preserve the original key type or cast consistently in both places.

To reproduce, see the new added testcase [`test_compute_handles_non_string_experiment_name`](https://github.com/instadeepai/winnow/pull/188/changes/38a302b557eee96c6021878611d3605b4f5f98fa).

File: winnow/calibration/calibration_features.py [hunk 2]
```diff
@@ -1330,3 +1374,61 @@ def compute(self, dataset: CalibrationDataset) -> None:
...
+        return train_data
+
+    def __getstate__(self) -> dict:
+        """Exclude transient regressor state from pickle serialisation."""
+        state = self.__dict__.copy()
+        state.pop("irt_predictors", None)
+        state.pop("irt_predictor", None)
+        return state
+
+    def __setstate__(self, state: dict) -> None:
```
Comment (BioGeek) [inline winnow/calibration/calibration_features.py]: When loading a `calibrator.pkl` produced before this change, the pickled `RetentionTimeFeature ` state has no `min_train_points`. `__setstate__ ` only restores the old state and creates  `irt_predictors`, so prediction with an existing/pretrained calibrator reaches  `self.min_train_points` in `_select_training_data` and raises `AttributeError`.

To reproduce, see the new test [`test_legacy_pickle_state_backfills_min_train_points`](https://github.com/instadeepai/winnow/pull/188/changes/91555866c6f36a2c33793d9795d701fdf9b2eb4a)

**Review by BioGeek — CHANGES_REQUESTED** [review]

File: winnow/calibration/calibration_features.py [hunk 3]
```diff
@@ -1209,65 +1195,116 @@ def check_valid_irt_prediction(self, dataset: CalibrationDataset) -> pd.Series:
...
+                    stacklevel=2,
+                )
+                experiments_to_fit = {"__global__": dataset.metadata}
+            else:
+                return
+        else:
+            experiments_to_fit = {
+                str(exp_name): group
+                for exp_name, group in dataset.metadata.groupby("experiment_name")
+                if exp_name not in self.irt_predictors
```
Comment (BioGeek) [inline winnow/calibration/calibration_features.py]: When experiment_name values are non-string (which parquet/IPC inputs can contain), saved regressors are keyed by `str(exp_name)` but this guard checks the raw `exp_name` value. A preloaded regressor under key '1' is treated as missing for experiment 1, so prediction refits it (or raises on insufficient inference data) instead of using the configured checkpoint.

File: winnow/calibration/calibration_features.py [hunk 4]
```diff
@@ -1209,65 +1195,116 @@ def check_valid_irt_prediction(self, dataset: CalibrationDataset) -> pd.Series:
...
+                    "RT->iRT regressor. For multi-experiment data, ensure each "
+                    "spectrum file includes an 'experiment_name' column or use "
+                    "MGF format (which derives it from the filename).",
+                    stacklevel=2,
+                )
+                experiments_to_fit = {"__global__": dataset.metadata}
+            else:
+                return
+        else:
+            experiments_to_fit = {
```
Comment (BioGeek) [inline winnow/calibration/calibration_features.py]: When a `RetentionTimeFeature`/`calibrator` instance is reused for another dataset with the same `experiment_name`, this check skips fitting because `irt_predictors ` was populated by the previous `prepare()` call, even though those regressors were not loaded from a checkpoint. The subsequent `compute()` then uses the stale RT->iRT mapping, so repeated `calibrator.predict()`/`compute_features()` calls in the same process can silently produce incorrect iRT features.

**Review by BioGeek — CHANGES_REQUESTED** [review]

File: winnow/calibration/calibration_features.py (hunk 3 above)
Comment (JemmaLDaniel) [reply winnow/calibration/calibration_features.py]: I've stringified the raw `exp_name` :)

File: winnow/calibration/calibration_features.py (hunk 4 above)
Comment (JemmaLDaniel) [reply winnow/calibration/calibration_features.py]: I altered the logic to stop using `irt_predictors` membership as the skip signal and track which experiments were intentionally pinned from a checkpoint via `_loaded_experiment_names`. This way, we should be able to differentiate between regressors saved with a model checkpoint (no refitting) and regressors fitted in-memory (should be refitted if we see the same experiment name again)

**Review by BioGeek — APPROVED** [review]
Looks good to me now, thanks!
